test/Prototypes/Algorithms.swift.gyb

```diff
@@ -252,21 +195,24 @@ internal func _gcd(_ m: Int, _ n: Int) -> Int {
   return m
 }
 
-extension MutableCollection where Self: RandomAccessCollection,
-  SubSequence: MutableCollection, SubSequence: RandomAccessCollection {
+extension MutableCollection where Self: RandomAccessCollection {
 
   /// Rotates elements through a cycle, using `sourceForIndex` to generate
   /// the source index for each movement.
   @inline(__always)
-  internal mutating func _rotateCycle(start: Index,
-    transform sourceForIndex: (Index) -> Index)
+  internal mutating func _rotateCycle(
+    start: Index,
+    pivot: Index,
+    plus: IndexDistance,
+    minus: IndexDistance)
```

Comment:
FWIW, you could factor a wee bit more and replace `pivot`, `plus`, and `minus` with a closure:

```swift
  internal mutating func _rotateCycle(
    start: Index,
    sourceOffsetForIndex: (Index) -> IndexDistance
  ) {
    let tmp = self[start]
    var i = start, j = index(i, offsetBy: sourceOffsetForIndex(i))
    while j != start {
      self[i] = self[j]
      i = j
      j = index(j, offsetBy: sourceOffsetForIndex(j))
    }
    self[i] = tmp
  }
```

Reply:
Thanks! I missed this intermediate version when I was exclusivizing the memory access.